Fix sqlite timestamps

Sytten opened the pull request:
Fixes #3693 
Fixes #3320

This is a big change in the way we store and parse timestamp in sqlite.
- **(Major)** We don't ignore timezones anymore if they are present, I am unsure why this what put in place but it feels very very very wrong to me. The data written by diesel is always switched to UTC, but we can't assume the data in the database was written by us. We still fallback to a "naive" implementation if the timezone is missing. I had to remove "bad" tests since `1970-01-01 00:00+01:00` is not the same time as `1970-01-01 00:00Z` in my book. We still ignore the timezone if the user uses a "naive" datetime to read the data (I think that is fair, though we could verify with other ORM if they correct to UTC instead).
- Uniformity between time and chrono, the code looks more similar and the formats are the same so a user can switch library without any issue now (that was something I found in #3693)
- Fixed some documentation

Sytten commented on diesel/src/sqlite/types/date_and_time/chrono.rs:
```diff
@@ -7,21 +9,85 @@ use crate::serialize::{self, IsNull, Output, ToSql};
+const ENCODE_NAIVE_DATETIME_FORMAT_SUBSECOND: &str = "%F %T%.f";
+
+const ENCODE_DATETIME_FORMAT_WHOLE_SECOND: &str = "%F %T%:z";
+const ENCODE_DATETIME_FORMAT_SUBSECOND: &str = "%F %T%.f%:z";
```
I need to change `%.f` to `%.3f` here to be consistent with `time` crate

Sytten commented on diesel/src/sqlite/types/date_and_time/chrono.rs:
```diff
@@ -150,14 +178,33 @@ impl FromSql<TimestamptzSqlite, Sqlite> for NaiveDateTime {
 impl ToSql<TimestamptzSqlite, Sqlite> for NaiveDateTime {
     fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
-        out.set_value(self.format("%F %T%.f+00:00").to_string());
+        let format = if self.nanosecond() == 0 {
```
And change here to only care about the micro seconds to match the `%.3f`

weiznich commented on diesel/src/sql_types/mod.rs:
```diff
@@ -332,14 +332,12 @@ pub struct Time;
 /// - [`std::time::SystemTime`][SystemTime] (PG only)
 /// - [`chrono::NaiveDateTime`][NaiveDateTime] with `feature = "chrono"`
 /// - [`time::PrimitiveDateTime`] with `feature = "time"`
-/// - [`time::OffsetDateTime`] with `feature = "time"`
```
There are such impls for the mysql backend. Probably we should change it to (mysql only) instead of removing these from the list

weiznich commented on diesel/src/sqlite/types/date_and_time/chrono.rs:
```diff
@@ -636,22 +704,6 @@ mod tests {
             "1970-01-01T00:00:00+00:00",
             "1970-01-01T00:00:00.000+00:00",
             "1970-01-01T00:00:00.000000+00:00",
-            "1970-01-01 00:00+01:00",
```
I assume these are removed because the serialization/deserialization of these values changed? Can you explain how and why they changed?

(We need to be very careful here to not introduce a breaking change for existing users)

Sytten replied:
Agreed we need to be super careful, I consider this test to be a bug but it might be considered a breaking change depending on how you look at it.

Basically right now if you insert manually a timestamp with a timezone in sqlite and **read** it using diesel it will ignore the timestamp. So for example `1970-01-01T00:00:00-01:00` will be treated as if it is `1970-01-01T00:00:00Z` when it is in fact ` 1970-01-01T00:01:00Z`. So for me this is a bug, but it does mean that people might get different dates.

Why it works if you **write** with diesel is because it will always translate to UTC before writing it. I didnt change that behaviour (but we might want to).

weiznich replied:
I think we can argue here that this is a bug fix. We should add it as potential breaking change to the changelog although.

weiznich commented on diesel/src/sqlite/types/date_and_time/chrono.rs:
```diff
@@ -7,21 +9,85 @@ use crate::serialize::{self, IsNull, Output, ToSql};
 use crate::sqlite::Sqlite;
 
-const SQLITE_DATE_FORMAT: &str = "%F";
+const DATE_FORMAT: &str = "%F";
```
Can we add a comment here that these formats need to be in sync with that one in the `time` support module (and vice versa)?

Mingun replied:
Even better to make this comment doccomment and put a direct intradoc reference to the corresponding constant(?), then editors, like VS Code, will able to create a clickable link to it

weiznich reviewed (commented):
Thanks for opening this PR. I've left a bunch of questions and notes for minor improvements. Beside of that I want to clarify an important point: You consider that a minor change, as it does only strictly extend which time formats are parsed and does not error out on formats that worked before?

Additionally I would like to see a few tests for time<->chrono roundtrips just to make sure that they stay consistent in the future.

Sytten commented:
@weiznich This is ready for another round of review, I changed a bit the encoding.

I realized we didn't need the differentiation for `subsecond` in `chrono` since the dot is handled by the formatter.
I changed `time` to now print all digits instead of just 3 or 6 like it currently done, this aligns it with `chrono` (and what `sqlx` is doing whatever that's worth).

I am still unsure how many digits we should really print, PG and Mysql cut at the microseconds but sqlite by default stops at the millisecond with `%f`. And if we should differentiate between a datetime with timezone and a "naive" datetime/time. Any external opinion would be appreciated, but it seems all over the place.

The case for printing they all is quite simple, it's still respecting the RFC3339 and the user can decide to not pass us the microseconds/nanoseconds if they don't want them in the database but the inverse (stopping at X digits) means we decide what loss of information is acceptable.

weiznich commented on diesel/src/sqlite/types/date_and_time/chrono.rs:
```diff
@@ -7,21 +8,89 @@ use crate::serialize::{self, IsNull, Output, ToSql};
 use crate::sqlite::Sqlite;
 
-const SQLITE_DATE_FORMAT: &str = "%F";
+// Warning to future editors:
+// Changes in the following formats need to be kept in sync
+// with the formats of the "time" module.
+// We do not need a distinction between whole second and
+// subsecond since %.f will only print the dot if needed.
+// We always print as many subsecond as his given to us,
+// this means the subsecond part can be 3, 6 or 9 digits.
```
```suggestion
/// Warning to future editors:
/// Changes in the following formats need to be kept in sync
/// with the formats of the ["time"](super::time) module.
/// We do not need a distinction between whole second and
/// subsecond since %.f will only print the dot if needed.
/// We always print as many subsecond as his given to us,
/// this means the subsecond part can be 3, 6 or 9 digits.
```

[as suggested here: ](https://github.com/diesel-rs/diesel/pull/3694/files#r1275723428)

weiznich commented on diesel/src/sqlite/types/date_and_time/mod.rs:
```diff
@@ -92,3 +92,265 @@ impl ToSql<sql_types::TimestamptzSqlite, Sqlite> for String {
+    }
+
+    #[test]
+    fn time_to_chrono_date() {
```
Can we actually check the results of that queries for whether or not they return the "right" values?

weiznich commented on diesel/src/sqlite/types/date_and_time/time.rs:
```diff
@@ -14,23 +14,30 @@ use crate::serialize::{self, IsNull, Output, ToSql};
 use crate::sql_types::{Date, Time, Timestamp, TimestamptzSqlite};
 use crate::sqlite::Sqlite;
 
+// Warning to future editors:
+// Changes in the following formats need to be kept in sync
+// with the formats of the "chrono" module.
+// We need a distinction between whole second and subsecond
+// since there is no format option to forgo the dot.
+// We always print as many subsecond as his given to us,
+// this means the subsecond part can be between 1 and 9 digits.
```
```suggestion
/// Warning to future editors:
/// Changes in the following formats need to be kept in sync
/// with the formats of the ["chrono"](super::chrono) module.
/// We need a distinction between whole second and subsecond
/// since there is no format option to forgo the dot.
/// We always print as many subsecond as his given to us,
/// this means the subsecond part can be between 1 and 9 digits.
```

weiznich reviewed (commented):
Thanks for the update and sorry for taking so long to do the review.

I think it's mostly fine now. I would like to change the comments to doc-comments as suggested my @Mingun 
I also would like to actually check the values in the compatibility tests to ensure that we load the right thing there.

Sytten commented:
@weiznich Ready!

weiznich reviewed (approved):
Thanks :+1:
